### PR TITLE
[IMP] base, http_routing, website: allow custom routing rule 

### PR DESCRIPTION
**Pull request by JKE-be**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by JKE-be** [comment]
@KangOl 

Friendly reminder :)

File: addons/website/models/ir_http.py [hunk 1]
```diff
@@ -62,6 +64,57 @@ def get_request_website():
...
+    @classmethod
+    def _match(cls, path_info, key=None):
+        key = key or (request and request.website_routing)
+        return super(Http, cls)._match(path_info, key=key)
+
+    @classmethod
+    def _generate_routing_rules(cls, modules, converters):
+        website_id = request.website_routing
+        logger.debug("_generate_routing_rules for website: %s", website_id)
+        domain = ['|', ('redirect_type', '=', 'rewrite'), ('redirect_type', '=', 'not_found'), '|', ('website_id', '=', False), ('website_id', '=', website_id)]
```
Comment (KangOl) [inline addons/website/models/ir_http.py]:
```suggestion
        domain = [('redirect_type', 'in', ('rewrite', 'not_found')), '|', ('website_id', '=', False), ('website_id', '=', website_id)]
```

File: addons/website/models/ir_http.py [hunk 2]
```diff
@@ -62,6 +64,57 @@ def get_request_website():
...
+
+        rewrites = dict([(x.url_from, x) for x in request.env['website.rewrite'].sudo().search(domain)])
+
+        for u, e, r in super(Http, cls)._generate_routing_rules(modules, converters):
+            r = dict(r)
+            if u in rewrites:
+                rewrite = rewrites[u]
+                url_to = rewrite.url_to
+                if rewrite.redirect_type == 'rewrite' and u != url_to:
```
Comment (KangOl) [inline addons/website/models/ir_http.py]: When `u == url_to`, the url is ignored and act as a 404. It should be yield (once)

File: addons/website/models/ir_http.py [hunk 3]
```diff
@@ -174,8 +228,12 @@ def _serve_page(cls):
     @classmethod
     def _serve_redirect(cls):
         req_page = request.httprequest.path
-        domain = [('url_from', '=', req_page)] + request.website.website_domain()
-        return request.env['website.redirect'].search(domain, limit=1)
+        domain = [
+            '|', ('redirect_type', '=', 'redirect_301'), ('redirect_type', '=', 'redirect_302'),
```
Comment (KangOl) [inline addons/website/models/ir_http.py]:
```suggestion
            ('redirect_type', 'in', ('redirect_301', 'redirect_302')),
```

File: odoo/addons/base/models/ir_http.py [hunk 4]
```diff
@@ -245,24 +245,36 @@ def _postprocess_args(cls, arguments, rule):
...
+            mods = sorted(installed)
             # Note : when routing map is generated, we put it on the class `cls`
             # to make it available for all instance. Since `env` create an new instance
             # of the model, each instance will regenared its own routing map and thus
             # regenerate its EndPoint. The routing map should be static.
-            cls._routing_map = http.routing_map(mods, False, converters=cls._get_converters())
-        return cls._routing_map
+            routing_map = werkzeug.routing.Map(strict_slashes=False, converters=cls._get_converters())
+            for u, e, r in cls._generate_routing_rules(mods, converters=cls._get_converters()):
+                xtra_keys = 'defaults subdomain build_only strict_slashes redirect_to alias host'.split()
```
Comment (KangOl) [inline odoo/addons/base/models/ir_http.py]: Missing `methods` in accepted keys

File: odoo/addons/base/models/ir_http.py [hunk 5]
```diff
@@ -87,8 +87,8 @@ def _get_converters(cls):
         return {'model': ModelConverter, 'models': ModelsConverter, 'int': SignedIntConverter}
 
     @classmethod
-    def _find_handler(cls, return_rule=False):
-        return cls.routing_map().bind_to_environ(request.httprequest.environ).match(return_rule=return_rule)
+    def _match(cls, path_info, key=None):
+        return cls.routing_map(key=key).bind('').match(return_rule=True, method='POST', path_info=path_info)
```
Comment (KangOl) [inline odoo/addons/base/models/ir_http.py]: Why only match `POST` routes?

File: addons/website/models/website_rewrite.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    _name = 'website.route'
+    _description = "All Website Route"
+
+    path = fields.Char('Route')
+
+    @api.model
+    def _name_search(self, name='', args=None, operator='ilike', limit=100, name_get_uid=None):
+        res = super(WebsiteRoute, self)._name_search(name=name, args=args, operator=operator, limit=limit, name_get_uid=name_get_uid)
+        if not len(res):
+            self.env['website.rewrite']._refresh()
```
Comment (odony) [inline addons/website/models/website_rewrite.py]: why not call directly `self._refresh()` here?

File: addons/website/models/website_rewrite.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        res = super(WebsiteRewrite, self).unlink()
+        self._invalidate_routing()
+        return res
+
+    def _invalidate_routing(self):
+        # call clear_caches on this worker to reload routing table
+        self.env['ir.http'].clear_caches()
+
+    def refresh_routes(self):
+        if self.env['website.route'].check_access_rights('write'):
```
Comment (odony) [inline addons/website/models/website_rewrite.py]: there should be no need to explicitly check access here, as `website.route._refresh()` won't be able to create/delete routes if the user has no access, right?

+ add ACLs for `website.route`

Comment (JKE-be) [reply addons/website/models/website_rewrite.py]: my bad, lost during the morning rebase with all conflicts: https://github.com/odoo/odoo/compare/b561ee507f0d8ebcdd46197a49e268873efbc07f..18e81abe80e0d4c91c6c057ca56534dfef5bdbeb#diff-2516dc5a739891abf193cd9902befc65L19

File: addons/http_routing/models/ir_http.py [hunk 8]
```diff
@@ -168,6 +169,40 @@ def url_for(path_or_uri, lang_code=None):
...
+            else:
+                path = url_from
+            # TODO: fix mockRequest to make test successful with ._match()
+            #_ = request.registry['ir.http']._match(path)
+            _ = request.httprequest.app.get_db_router(request.db).bind('').match(path, method='POST')
+        except werkzeug.routing.RequestRedirect as e:
+            # remove query string from current env
+            new_url = e.new_url.split('?')[0]
+            # remove scheme and add query_string from url_from
+            new_url = new_url[7:] + (qs and '?%s' % qs or '')
```
[the remark this answers is not in the document]
Comment (JKE-be) [reply addons/http_routing/models/ir_http.py]: > bind(server_name, script_name=None, subdomain=None, url_scheme='http', default_method='GET', path_info=None, query_args=None)

we use default url_scheme from werkzeug bind method => so http://

File: addons/website/models/ir_http.py (hunk 2 above)
Comment (JKE-be) [reply addons/website/models/ir_http.py]: :+1:

File: odoo/addons/base/models/ir_http.py (hunk 5 above)
Comment (JKE-be) [reply odoo/addons/base/models/ir_http.py]: basé sur cette vieille heuristique: 'https://github.com/odoo/odoo/blob/bd3a78847998d117ebb32b303eaa3886f00c50ba/addons/http_routing/models/ir_http.py#L171'

Mais qui ne semble pas / plus vrai...

Mais est ce que finalement on veut gérer le cas des routes GET only ? 
POST only c'est certain, mais GET Only ?


Pour le moment, en attendant, je ré-utilise la request.method

File: odoo/addons/base/models/ir_http.py (hunk 4 above)
Comment (JKE-be) [reply odoo/addons/base/models/ir_http.py]: added in routing_map.add as previously

well spotted :eye:

File: addons/http_routing/models/ir_http.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (KangOl) [reply addons/http_routing/models/ir_http.py]: Ok, it's not clear without knowing the signature of `bind` method. Maybe force the scheme in the call to `bind`?

**Comment by JKE-be** [comment]
@robodoo r+

**Comment by rdeodoo** [comment]
@robodoo retry

**Comment by rdeodoo** [comment]
Blablabla, saw that one coming

**Comment by odony** [comment]
@robodoo retry and be nice to @rdeodoo ;)

**Comment by qsm-odoo** [comment]
@robodoo retry

**Comment by qsm-odoo** [comment]
@robodoo likes me :)

File: addons/http_routing/models/ir_http.py [hunk 9]
```diff
@@ -164,10 +165,46 @@ def url_for(path_or_uri, lang_code=None):
...
+
+        :param url_from: The URL to convert.
+        :param lang_code: Must be the lang `code`. It could also be something
+                          else, such as `'[lang]'` (used for url_return).
+        :param no_rewrite: don't try to match route with website.rewrite.
+    '''
+    new_url = False
+
+    # don't try to match route if we know that no rewrite has been loaded.
+    if not request.env['ir.http']._rewrite_len.get(request.website_routing):
```
Comment (Feyensv) [inline addons/http_routing/models/ir_http.py]: @JKE-be When website isn't installed, request.website_routing isn't initialized and this crashes.
Could you fix it? (If I do it, I'm just gonna c/p 2 lines of code from website to here :P)
https://github.com/odoo/odoo/issues/38020

`website/models/ir_http.py
`image

**Comment by go4site** [comment]
Much awaited feature... Thanks for adding features for custom URLs. 
By any mean, is it possible to have /custom-url  (Direct URL alias) instead of having /shop/custom-url-id for product URLs?
I have been waiting for such possibilites for years... I hope this one is available in Odoo 14.

**Comment by JKE-be** [comment]
@go4site 

unfortunately not in v14.

We changed the 
`/shop/product/product_name-id` to` /shop/product_name-id`
`/blog/post/blog_name-id` to `/blog/blog_name-id`

But we cannot just remove /shop or /blog, since it permit to define the model... without it, we just have the id of a record :/

**Comment by go4site** [comment]
Thanks for prompt response. In fact, we are using below module to achieve same results:

https://github.com/OCA/website/tree/10.0/website_seo_redirection

There is possibility to use any custom URL for example:

Defining https://example.com/amazing for https://example.com/blog/our-news-1/post/amazing-post-23

It also has feature of relcoating the controller itself. 

It would be great to have such feature in Odoo core. Actually, we need to upgrade couple of current deployments so it is mandatory for us to keep same URL structure. 

Would appreciate, if you please have a look on this module and see any possibility to add this in Odoo Core. Thanks.

**Comment by JKE-be** [comment]
Unfortunately it will be not possible before the v14 to change it.
But yes, we have some plan to change the routing in a future version to make it more customizable.
But not ETA at this moment.

**Comment by go4site** [comment]
My bad!!! Anyways, thanks :) 

By the way, for future reference it will be great to have a centralized user interface to manage all URLs. I really like Joomla's approach where URLs are centrally managed in Menu manager itself. So, you have full control of all menu items alongwith URls and publishing options etc. 

Joomla-ref

**Comment by JKE-be** [comment]
Remark, you can now in 14 use alias too instead of record diplsay_name.
image

It is not centralized for now, because we have 3 distinct model:

The page created at fly
The controller from python code
And the redirection/Rewrite that try to add a transparent layer to make it more customizable.

But as explain previously it is not impossible that redirect table and controller will be regouped in only one table in the future. We have some poc for it.

(In practice, we have one 4th table with controller already in db, but only used for autocomplete in rewrite and not to customize it, purpose is to merge all these one)

**Comment by go4site** [comment]
Yes, I already tested this feature. It is a great option. It would be great if you make this field available at the backend to support bulk import/export. 

It provides felxibility to have a custom url insteal of name slug. However,  it still has route prefix (/shop) in this case and id (9) in the URL.

**Comment by go4site** [comment]
It is not centralized for now, because we have 3 distinct model:

The page created at fly
The controller from python code
And the redirection/Rewrite that try to add a transparent layer to make it more customizable.

But as explain previously it is not impossible that redirect table and controller will be regouped in only one table in the future. We have some poc for it.

(In practice, we have one 4th table with controller already in db, but only used for autocomplete in rewrite and not to customize it, purpose is to merge all these one)

Yes, grouping controller and redirect will help to streamline URL handling. In addition, an internal router could be used to map final URL and internal http route per record basis. This final URL field (URL alias) could be made available in menu manager (central interface) instead of setting on individual pages, products, blogs etc. Also, it will be good to have sitemap inclusion option against each menu item.

**Comment by go4site** [comment]
Update:

I was able to get rid of /shop using 308 rewrite as below. However, "/shop" is hardcoded in the code:

def _compute_website_url(self):
        super(ProductTemplate, self)._compute_website_url()
        for product in self:
            if product.id:
                product.website_url = "**/shop/**%s" % slug(product)

So, it is breaking the change on the website. Is there any way to make it dynamic to adopt final URls as per rewrite rule?


prod-url


Screenshot_2020-09-03 308 - shop-prod - Odoo

**Comment by JKE-be** [comment]
Let's continue discussion in another thread (our forum e.g.) to avoid spam others employees.

But basically 308 will automatically redirect old one to the new one. So end user will never see /shop inurl
And in most of case (when you are not connected as editor) url are replaced on the fly by the url_for.

**Pull request closed without merging**